File: java/test/sorter/SorterTest.java
package sorter;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class SorterTest {

    @Test
    void testSmallSort() {
        
        int recordSize = 100;
        int recordCount = 10;
        
        String test = "testSmallSort: ";
        
        Provider generator = new RandomGenerator(recordCount, recordSize);
        Printer printer1 = new Printer(generator, test+"from generator");
        Witness lower = new Witness(printer1);
        Sorter sorter = new Sorter(new SorterConfig(), lower, recordSize);
        Witness upper = new Witness(sorter);
        Printer printer2 = new Printer(upper, test+"from sorter");
        Consumer consumer = new Consumer(printer2);
        consumer.consume();
        
        assertEquals(recordCount, lower.getCount(), "The count of the lower witness was " + lower.getCount() + " but should have been 10");
        assertEquals(recordCount, upper.getCount(), "The count of the upper witness was " + upper.getCount() + " but should have been 10");
        assertFalse(lower.isSorted(),  "The lower witness should not have been sorted but was");
        assertTrue(upper.isSorted(), "The upper witness was not sorted but should have been");
        assertEquals(lower.getCrc(), upper.getCrc(), "The checksum of the lower witness did not equal the checksum of the upper but should have");
    }

    @Test
    void testAllMemory() {
        
        int recordSize = 1000;
        int recordCount = 90000;
        
        @SuppressWarnings("unused")
        String test = "testAllMemory: ";
        
        Provider generator = new RandomGenerator(recordCount, recordSize);
        Witness lower = new Witness(generator);
        Sorter sorter = new Sorter(new SorterConfig(), lower, recordSize);
        Witness upper = new Witness(sorter);
        Consumer consumer = new Consumer(upper);
        consumer.consume();
        
        assertEquals(recordCount, lower.getCount(), "The count of the lower witness was " + lower.getCount() + " but should have been "+recordCount);
        assertEquals(recordCount, upper.getCount(), "The count of the upper witness was " + upper.getCount() + " but should have been "+recordCount);
        assertFalse(lower.isSorted(),  "The lower witness should not have been sorted but was");
        assertTrue(upper.isSorted(), "The upper witness was not sorted but should have been");
        assertEquals(lower.getCrc(), upper.getCrc(), "The checksum of the lower witness did not equal the checksum of the upper but should have");
    }

    @Test
    void testSpillToSsdFewBlocks() {
        
        int recordSize = 1024;
        int recordCount = 1024*100;
        
        @SuppressWarnings("unused")
        String test = "testSpillToSsdFewBlocks: ";
        
        Provider generator = new RandomGenerator(recordCount, recordSize);
        Witness lower = new Witness(generator);
        Sorter sorter = new Sorter(new SorterConfig(), lower, recordSize);
        Witness upper = new Witness(sorter);
        Consumer consumer = new Consumer(upper);
        consumer.consume();
        
        sorter.printStats();
        
        assertEquals(recordCount, lower.getCount(), "The count of the lower witness was " + lower.getCount() + " but should have been "+recordCount);
        assertEquals(recordCount, upper.getCount(), "The count of the upper witness was " + upper.getCount() + " but should have been "+recordCount);
        assertEquals(lower.getCrc(), upper.getCrc(), "The checksum of the lower witness did not equal the checksum of the upper but should have");
        assertFalse(lower.isSorted(),  "The lower witness should not have been sorted but was");
        assertTrue(upper.isSorted(), "The upper witness was not sorted but should have been");
    }


    @Test
    void testSpillToSsd() {
        
        int recordSize = 1000;
        int recordCount = 900000;
        
        @SuppressWarnings("unused")
        String test = "testSpillToSsd: ";
        
        Provider generator = new RandomGenerator(recordCount, recordSize);
        Witness lower = new Witness(generator);
        Sorter sorter = new Sorter(new SorterConfig(), lower, recordSize);
        Witness upper = new Witness(sorter);
        Consumer consumer = new Consumer(upper);
        consumer.consume();
        
        sorter.printStats();
        
        assertEquals(recordCount, lower.getCount(), "The count of the lower witness was " + lower.getCount() + " but should have been "+recordCount);
        assertEquals(recordCount, upper.getCount(), "The count of the upper witness was " + upper.getCount() + " but should have been "+recordCount);
        assertEquals(lower.getCrc(), upper.getCrc(), "The checksum of the lower witness did not equal the checksum of the upper but should have");
        assertFalse(lower.isSorted(),  "The lower witness should not have been sorted but was");
        assertTrue(upper.isSorted(), "The upper witness was not sorted but should have been");
    }

    @Test
    void testSpillToHdd() {
        
        int recordSize = 1000;
        int recordCount = 19000;
        
        @SuppressWarnings("unused")
        String test = "testSpillToSsd: ";
        
        SorterConfig cfg = new SorterConfig();
        cfg.ssdStorageSize = 0*1024*1024;
        cfg.memoryBlockCount = 10;
        
        Provider generator = new RandomGenerator(recordCount, recordSize);
        Witness lower = new Witness(generator);
        Sorter sorter = new Sorter(cfg, lower, recordSize);
        Witness upper = new Witness(sorter);
        Consumer consumer = new Consumer(upper);
        consumer.consume();
        
        sorter.printStats();
        
        assertEquals(recordCount, lower.getCount(), "The count of the lower witness was " + lower.getCount() + " but should have been "+recordCount);
        assertEquals(recordCount, upper.getCount(), "The count of the upper witness was " + upper.getCount() + " but should have been "+recordCount);
        assertEquals(lower.getCrc(), upper.getCrc(), "The checksum of the lower witness did not equal the checksum of the upper but should have");
        assertFalse(lower.isSorted(),  "The lower witness should not have been sorted but was");
        assertTrue(upper.isSorted(), "The upper witness was not sorted but should have been");
    }

}

File: Testing/Tests/Sorters/SorterTest.cpp
#include "SorterTest.h"

void SorterTest::testSmallSort() {
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 10;
    RandomGenerator rng(*cfg);

    string test = "testSmallSort: ";

    Printer printer1(&rng, test+"from generator");
    Witness lower(&printer1);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Printer printer2(&upper, test+"from sorter");
    Consumer consumer(&printer2);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testAllMemory() {

    string test = "testAllMemory: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 90000;
    cfg->recordSize = 1000;
    RandomGenerator rng(*cfg);
    Witness lower(&rng);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testSpillToSsdFewBlocks() {
    string test = "testSpillToSsdFewBlocks: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 1024*100;
    cfg->recordSize = 1024;
    RandomGenerator rng(*cfg);
    Witness lower(&rng);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testSpillToSsd() {
    string test = "testSpillToSsd: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 900000;
    cfg->recordSize = 1000;
    RandomGenerator rng(*cfg);
    Witness lower(&rng);
    SorterConfig* cfg2 = new SorterConfig();
    cfg2->recordCount = 900000;
    cfg2->recordSize = 1000;
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testSpillToHdd() {
    string test = "testSpillToLotsOfHddRuns: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 190000;
    cfg->recordSize = 1000;
    RandomGenerator rng(*cfg);
    Witness lower(&rng);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testSpillToLotsOfHddRuns() {
    string test = "testSpillToLotsOfHddRuns: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 190000;
    cfg->recordSize = 1000;
    cfg->memoryBlockCount = 10;
    cfg->memoryBlockSize = 100*1024*1024;
    RandomGenerator r(*cfg);
    Witness lower(&r);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

void SorterTest::testZeroRecords() {
    string test = "testZeroRecords: ";
    SorterConfig* cfg = new SorterConfig();
    cfg->recordCount = 0;
    cfg->recordSize = 1000;
    RandomGenerator rng(*cfg);
    Witness lower(&rng);
    SorterConfig* cfg2 = new SorterConfig();
    Sorter sorter(*cfg2, &lower);
    Witness upper(&sorter);
    Consumer consumer(&upper);
    consumer.consume();

    sorter.printStats();

    assert(("The count of the lower witness should have equaled 0" && 0 == lower.getCount()));
    assert(("The count of the upper witness should equaled 0" && 0 == upper.getCount()));
    assert(("The count of the lower witness should have equaled the record count" && cfg->recordCount == lower.getCount()));
    assert(("The count of the upper witness should equaled the record count" && cfg->recordCount == upper.getCount()));
    assert(("The checksum of the lower witness did not equal the checksum of the upper but should have" && lower.getCrc() == upper.getCrc()));
    assert(("The lower was sorted but should not have been" && !lower.isSorted));
    assert(("The upper witness was not sorted but should have been" && upper.isSorted));
}

int main(){
    SorterTest test;
    test.testSmallSort();
//    test.testAllMemory();
//    test.testSpillToSsdFewBlocks();
//    test.testSpillToSsd();
//    test.testSpillToHdd();
//    test.testSpillToLotsOfHddRuns();
//    test.testZeroRecords();
}